### PR TITLE
Rob10470/load windows data file

### PR DESCRIPTION
**Pull request by robert-craig-houston**
### Summary
Hi,  this is quite an ugly set of changes, but it establishes a workflow that I should be able to keep up with for however long until we get C++ 20. HH's library fits its purpose like a glove, which means a lot of macro usage and unhappy changes. 

Apologies for the wall of text, and thanks for the help! Oh and don't mind the line count changes, embedded files and all that!

This PR does the following:
- Switches loading mechanism to std::filesystem to enable windows to load compiled tzif files.
- Fixes a host of minor issues loading TZIF files in HH's library.
- Uses HH support for Posix Timezones to support minimized TZIF files Posix Footers. This is initially done via repeatedly parsing the Posix TZ to avoid larger changes to the library, but I hope to switch us over to directly storying the Posix Info inside the TZ.
- Adds functioning support for fetching sys_info (UTC info on epochs), and starts the process for Local timezones but this requires more testing. This is not a core usecase for either arcade or transport networks, and unless there is a strong fear this will inadvertantly become load bearing I might tackle this as a seperate issue.
- Adds an embedded data file template, and a python script capable of embedding TZIF information inside it to form a CPP file which can be built into the binary. 
- Adds the product of this embedding, produce with TZDB-2022c. This is NOT the most up to date TZDB available, and will need to be updated, an issue has been raised in Arcade so I can get myself on notification lists for these updates.
- Leaves as much of the code intact as reasonable so we can switch back to loading from uncompiled TZDB's without too much difficulty.

There is a sister PR in RTC that tests all of this. This is done via an exhaustive generation of all UTC transition epochs until 2032 stored in a new JSON file, and a host of other tests. Unfortunately there is a third party architecture depdendencies file that seems necessary to alter but its not in a git repo, and I don't know where it is fetched from!

### Future Work
This has churned up some issues about edge cases in the library. I presume user error for most of these and non are insurmountable. See our epic here: https://devtopia.esri.com/runtime/arcade/issues/779, for all of them, but the relevant ones to third party updates

1. TZIF Posix footers have small extensions to the Posix format, two timezones fail due to this
2. Zic appears to be incorrectly compiling certain TZIF files for America/Peurto_Rico and US/Virgin, ignoring WW2 transitions between war and peace. Tzif files are easy to generate so if the problem persists on update I'll consider writing a script to write them ourselves.
3. HH's library appears to be outputting incorrect units for some unusual timezones such as WET via the uncompiled TZDB. Where it should be returning seconds it returns milliseconds, unfortunately this breaks my testing strategy as on the way back in it reads the epoch as 1000x further than it is.
4. Arcade should move its bespoke TZ handling into HH, it gives us a better language for expressing how we behave in edge cases. A good reason to introduce heavy testing for the Local TZ math codepaths.
5. Working out TZIF files by hand after hex dumping them is a pain. I'd like to write a script that dumps it formatted to a file, along with relevant parts of the JSON tests and uncompiled TZDB. This would be quite a bit of work so I'm leaving it open until I hit my head on that problem again.

### Current Workflow:
I've not yet written this up in detail, I'd rather hold off on that until I'm certain everything is ticking, but as an overview:

1. Pull our 3rd party data library and do our regular updates to main and RTC as typical with 3rd party libraries if possible. If not due to merge conflicts we may need to get selective. The library isn't suffering huge churn so this isn't a disaster.
2. Check if we need to update the [IANA TZDB](https://www.iana.org/time-zones). Some updates are relatively minor and its possible Pro and RTC won't want to undergo the update process. If we don't, finish.
3. Pull the approriate IANA TZDB and code into a Linux Distro and compile the Zic Compiler. I've found this to be far more friendly in Linux and WSL than on windows, Hooray for make files. 
4. Run ZIC on all the TZDB files in the library and store in a location now called CompiledTZDB or some such. Be careful, not specifying a location might make Zic update your Linux Distro's TZDB.
5. Copy the `leapseconds` file from the uncompiled TZDB to the compiled tzdb.
6. Copy sweet delicious [windows timezone information.](https://github.com/unicode-org/cldr/blob/main/common/supplemental/windowsZones.xml) into the compiled TZDB.
7. Run EmbedData.py PathToCompiledTZDB and copy the results into the src folder of date. 
8. Test everything compiled sucessfully, expect some test failures due to TZDB update.
9. Use the non-compiled tzdb with a clean version of HH's library to generate test data, and run against your compiled version. Debug or Celebrate.

### CODE REVIEW

File: date.lua
Comment (chri7325) [inline date.lua]: No need for this change.

File: embedded_data/embed_data.py
Comment (chri7325) [inline embedded_data/embed_data.py]: missing endl

File: embedded_data/embedded_file_template.cpp
Comment (chri7325) [inline embedded_data/embedded_file_template.cpp]: missing endl

File: embedded_data/embedded_file_template.cpp
Comment (chri7325) [inline embedded_data/embedded_file_template.cpp]: Formatting looks different on this file

File: src/tz.cpp
Comment (chri7325) [inline src/tz.cpp]: Is this an artifact?

File: src/tz.cpp
Comment (chri7325) [inline src/tz.cpp:95]: Does it hurt to just include this?

File: src/tz.cpp
Comment (chri7325) [inline src/tz.cpp]: Better to just leave this in

**Review by chri7325 — COMMENTED** [review]
I can't really comment on the actual code changes since I don't know much about how this works but I would like to comment on the long term maintenance of these changes.

The commits on this PR seem like intermediate commits than can all be redone. I would recommend to make any commits that just add things like the embedded data as a single commit with all the details on how they were generated and added for good history.

After that, the commits that are actually made to the source code can be a separate commit. Doing it this way will make it a lot easier to update the 3rdparty library as any potential merge conflicts can be reverted, then the library updates be merged cleanly, followed by a reapplication of the changes to the source code. Any new generated files won't need to be backed out because there's no way any conflicts can arise from that so it is important to separate these to make future updates much easier.

Lastly, I see a lot of changes to the source code as potentially unnecessary. The less changes we make to the upstream code means the less pain it will be to update and deal with conflicts so every change should be the minimal set possible. For example, instead of changing a function signature or deleted code that won't be used, just short circuit it with an early return of our code. This may make the code a little harder to understand but we can always use comments to let a reader know that we're doing that. Keeping all our our edits.

File: src/tz.cpp
Comment (robert-craig-houston) [reply src/tz.cpp]: Not entirely, leaving this in triggers some of the other #ifdef blocks on loading. Given your above comments I'll move this to the .Lua file to avoid extra code changes.

File: embedded_data/embedded_file_template.cpp
Comment (robert-craig-houston) [reply embedded_data/embedded_file_template.cpp]: I've run format.sh on this file to bring it inline with RTC, although its reporting no changes. Fixed some minor spacing issues, manually, but am struggling to spot other formatting issues. Sorry if I'm being blind, could you highlight some of them in case I'm being daft!

File: src/tz.cpp
Comment (robert-craig-houston) [reply src/tz.cpp:95]: It caused a failure to build on windows if included. Replaced usages with filesystem for debug testing then removed entirely. I"ll replace this deletion with a commented out include with info to minimize diff change, unless you have a style preference?

HH is using the fact that windows doesnt have a OS TZDB available to make this work.

File: src/tz.cpp
Comment (robert-craig-houston) [reply src/tz.cpp]: Re-examining this change I can leave this in but it has knock on effects to a lower section of code responsible for fetching the TZDIR. That section can have its ifdef set to false, which does reduce over all line changes.

**Comment by robert-craig-houston** [comment]
Thanks for the insight @chri7325!

I'll take another crack at this with your suggestions and try to minimise line changes in the diff. The next draft PR will also probably include some messy history for my benefit, I'll then fix the history of once the diff is signed off on.

As you say for the actual mergeable PRs- Seperate the embedded data first because it won't trigger merge conflicts.

Closing PR as branch now outdated

File: embedded_data/embedded_file_template.cpp
Comment (chri7325) [reply embedded_data/embedded_file_template.cpp]: Without a clang-format file at the root, the format.sh script won't be able to get it in a way where it follows RTC exactly. The most obvious one here is the use of tabs instead of spaces. No biggie though, I just thought I'd point it out while we're here since the tabs on GitHub are exaggerated.

File: src/tz.cpp
Comment (chri7325) [reply src/tz.cpp:95]: I would recommend something like

```
if USE_OS_TZDB && !RTC_WINDOWS_FAMILY
```

This serves two purposes. It limits the change to a single line that (hopefullly) won't be changed by upstream and also adds our preprocessor defines, alerting a developer reading the code that this is a change we've made. It also makes hunting down our modifications easier in searches of `RTC_`

**Review by chri7325 — COMMENTED** [review]
Apologies for all the comments but it should be worth it long term. I think it is okay to just rebase your branch and push back to this PR as well since there's a lot of very good information here and keeping it all in one place is not a bad thing.

File: embedded_data/embedded_file_template.cpp
Comment (robert-craig-houston) [reply embedded_data/embedded_file_template.cpp]: Ahh, that makes sense 🤦 

Will do some fiddling to make sure the template gets properly formatted by the script. Thanks for the catch!

File: src/tz.cpp
Comment (robert-craig-houston) [reply src/tz.cpp:95]: Gotcha, in similar cases where I need to modify these macro blocks to always or never trigger do we have something like RTC_FALSE and RTC_TRUE?

Comment (chri7325) [reply src/tz.cpp:95]: These are the list of defines we define globally for all projects https://devtopia.esri.com/runtime/runtimecore/blob/main/docs/writing_code/patterns/preprocessor_defines.md. To get what you'd need, you could just check `if defined(RTC)` as a true for our code, false for other code. If you just need this for specific platforms, then you'd use the platform defines and if just for specific chipsets, you'd use the chipset defines. I hope with all those different combinations, you'd be able to tailor the build very specifically for every use case.

Comment (robert-craig-houston) [reply src/tz.cpp:95]: Thats awesome, thanks!

**Comment by robert-craig-houston** [comment]
Sorry for the delay @chri7325, was wrestling with a very obvious problem in reterospect that I don't know how to solve. 

For the library to work correctly the header files and the CPP files need to have the same defines because a lot of the logic is stored in the header files. If they aren't mysterious things start happening, like vectors mysteriously filling and emptying. However since the date.lua files defines don't effect RTC its not clear to me how to resolve this without polluting RTC projects. 

Presumably this isn't the first library to have this problem, was wondering what your thoughts were?

Thanks for the help and patience, both are very much appreciated! Also believe me when I say I'd rather more comments than less, always happy to take good advice rather than get burned 😆

**Comment by chri7325** [comment]
No worries, I love this stuff :). For defines that need to be added with consumers, I put them in the header file. Many libraries have a userconf.h file or some variation of that that is include in every header in the project. I'm hoping this library does as well in which case, you can put all of them there in a block. Some libraries don't have any define paths in the public header so for those, the defines just go into the lua file.

**Comment by chri7325** [comment]
Since everything is built on top of date.h, your best bet might be https://github.com/Esri/date/blob/master/include/date/date.h#L92 where you can add `// RTC Configuration` or something

**Comment by robert-craig-houston** [comment]
Gotcha, sounds good to me. I'll push the updated version sometime tommorow pre-commented for any unusual changes that should provide some more context for the bare necessary code changes. Hopefully with enough git-fu to keep the branch history intact.

Thanks for the help!

File: include/date/date.h
Comment (robert-craig-houston) [inline include/date/date.h]: This actually throws a warning about multiple defines. Wrapping it in IFDEF causes it to fail. Not sure what to make of it, and fixing it is a larger set of line changes.

File: include/date/tz.h
Comment (robert-craig-houston) [inline include/date/tz.h:789]: No getting around storing extra data here. HH gets around this by making assumptions about the TZIF format used by OS's that would increase the size of our embedded data.

File: include/date/tz.h
Comment (robert-craig-houston) [inline include/date/tz.h:1161]: We make use of the Windows TZ mapping code, so we need this field on all platforms

File: include/date/tz.h
Comment (robert-craig-houston) [inline include/date/tz.h:841]: With the change above (Including posix timezones) we also can no longer uniquely identify a timezone by its transition information from the TZIF file because the posix timezone might be in effect. 

Passing the timepoint in was the simplest solution I could find

File: include/date/tz.h
Comment (robert-craig-houston) [inline include/date/tz.h:1199]: This is an exposure of a previously static function to convert between windows and IANA timezones. Its previously been windows only but we'll want access to it for Transport Networks.

File: src/tz.cpp
Comment (robert-craig-houston) [inline src/tz.cpp:35]: If def'd these needed includes to make it obvious this is RTC configuration. These are the first non-copywrite lines in the file so the extra length seemed worth it.

File: src/tz.cpp
Comment (robert-craig-houston) [inline src/tz.cpp:499]: We need to use some code from the following section to handle windowsZones.xml, but not all of it. The !USE_OS_TZDB #if reappears below.

File: src/tz.cpp
Comment (robert-craig-houston) [inline src/tz.cpp:707]: Is it worth me updating these comments to include the defined(RTC) condition?

File: src/tz.cpp
Comment (robert-craig-houston) [inline src/tz.cpp:710]: This code begins including Uncompiled TZDB parsing code, which conflicts with our codepath

File: src/tz.cpp
Comment (robert-craig-houston) [inline src/tz.cpp:2131]: To reduce line changes define a lambda that handles fetching Posix Info correctly. Its an ugly one liner but just links the posix timezone and the final timezone transition together.

File: src/tz.cpp
Comment (robert-craig-houston) [inline src/tz.cpp:2135]: The TZIF standard says the last transition, once in effect, must be equivalent to the start of the Posix TZ footer. HH avoids having to handle this because OS's seem to pre-calculate their transitions.

File: src/tz.cpp
Comment (robert-craig-houston) [inline src/tz.cpp:2184]: This logic needs extensive testing, but is correct enough to pass mild scrutiny. We get a chicken and egg problem where to get the UTC timepoint we need to calculate the UTC timepoint.

This code isn't needed for either workflow *yet*, so I kept as simple as would compile.

File: src/tz.cpp
Comment (robert-craig-houston) [inline src/tz.cpp:2765]: I couldn't find two ways about it. This code relies on Dirent, which wasn't suitable for windows. I've kept the logic of the dataloading the same, but unfortunately needed to nix the loading logic.

On the brightside this area of code is unlikey to experience much churn,

**Comment by robert-craig-houston** [comment]
@chri7325, apologies for the wall of comments, in reterospect should have started a review! These are intended to annotate the files when viewing the diff.

Also recommend viewing all changes, not just those since last commit because I ran a revert on the branch back to runtimecore state to keep the history but let me re-apply edits.

Thanks for the help!

File: date.lua
Comment (chri7325) [inline date.lua]: Add spacing; These seem to already be built into the compiler. Are these needed at all?

```
<command line>:8:9: warning: '__ORDER_BIG_ENDIAN__' macro redefined [-Wmacro-redefined]
#define __ORDER_BIG_ENDIAN__ 0
        ^
<built-in>:86:9: note: previous definition is here
#define __ORDER_BIG_ENDIAN__ 4321
        ^
In file included from <built-in>:427:
<command line>:9:9: warning: '__ORDER_LITTLE_ENDIAN__' macro redefined [-Wmacro-redefined]
#define __ORDER_LITTLE_ENDIAN__ 1
        ^
<built-in>:85:9: note: previous definition is here
#define __ORDER_LITTLE_ENDIAN__ 1234
```

File: include/date/date.h
Comment (chri7325) [reply include/date/date.h]: Since warnings as errors isn't an issue for 3rdparty, warnings are not a big deal. You can also remove this warning with `#    undef USE_OS_TZDB` right before it.

File: src/tz.cpp
Comment (chri7325) [reply src/tz.cpp:707]: Probably not, reducing the potential for update conflicts is a bigger thing to go for.

File: src/tz.cpp
Comment (chri7325) [inline src/tz.cpp]: Can in be null? Better to just remove this line if so and you want to go down the codepath?

**Review by chri7325 — COMMENTED** [review]
Looking good? Just a few more minor questions. I can't get things to compile on macOS

```
include/date/tz.h:1162:25: error: no member named 'timezone_mapping' in namespace 'date::detail'
    std::vector<detail::timezone_mapping> mappings;
```

I see that function is only defined for Windows.

File: date.lua
Comment (robert-craig-houston) [reply date.lua]: Lacking these triggers a failure to build in my MSVC build.

I could move these to the Windows platform if thats prefereable?

File: src/tz.cpp
Comment (robert-craig-houston) [reply src/tz.cpp]: It can't be null, we'd have already errored by this point, but the syntax is valid, scratch one line edit!

File: date.lua
Comment (chri7325) [reply date.lua]: Ahh I see. I think so since clang seems to have these built-in with different values so I'm not sure what effects that would have. Where are these values defined in the upstream project for Windows?

**Review by chri7325 — APPROVED** [review]
Just a few more minor changes and I think this is good to go. No need to re-review

File: date.lua
Comment (robert-craig-houston) [reply date.lua]: Apologies, I'm not familiar with what "Upstream" means in this context. 

From what I cant tell MSVC doesn't easily provide a way to access Endian data. The values I use are just to force HH to compile for little endian-ness. We can likely use the numbers provided from Clang, if that answers the question at all 😄

Comment (chri7325) [reply date.lua]: Ahh that's right. There's no "upstream" way to build this because the original repo (upstream) this is forked off of doesn't even use these codepaths at all on Windows.  Better to be consistent with the clang values then. Can you also put a comment on why these need to be defined on Windows.

**Comment by chri7325** [comment]
Oh and last but not least, before merging, it would be good to squash these commits into the structure that we went over and then do a normal merge commit after that :)

**Comment by robert-craig-houston** [comment]
Gotcha, will look to sort this out for tommorow. Just a few more questions, thanks for all the help and patience!

- Previously I have been unable to coax Date into building via vtest. Is this because architecture/dependencies.lua needs to be update to include an `readProject(_3RDPARTY_DIR .. "/date/date.lua")` entry? I don't have write access to the architecture repository to do that unfortunately!
- I don't know of any way to squash a branch without altering its history pre-merge, maybe my late night git fu is failing me! The easiest thing that springs to mind is to open a new PR just for the embedded data files, and then squash-merge this PR as the code changes to keep the history above while seperating those concerns, assuming that works for you? 

Thanks again for all the patience!

**Comment by chri7325** [comment]
I've given you full access to the 3rdparty org so you should be able to push that branch in now and use it on vTest. That is the correct way to do it so feel free to add me to the PR that puts it up.

As for redoing the commits, you can use an interactive rebase or better yet, a soft reset. Doing something like `git reset --soft runtimecore` will drop all your commit history but keep your final changes in tact. From there, you can recommit you final changes in clean sets. Once things are looking good for you locally, force push and this PR will have the new commits. It is probably a good idea to start off first with a temporary local branch to get used to it :)

**Comment by robert-craig-houston** [comment]
Ah, quite happy with those git shenigans, I abuse my local branches often enough 😆  I thought doing that might cause github to lose track of the comments in the code, but they transferred perfectly, the more you know!

I've updated both commits titles, and the data file commit describes loosely how I generated the data file. 

We've got an issue logged for updating the TZDB data: https://devtopia.esri.com/runtime/arcade/issues/1073, during which I'll automate and test the workflow more thoroughly, unless you feel strongly otherwise?

Thanks for all the help!

**Review by chri7325 — APPROVED** [review]
Perfect!

**Comment by chri7325** [comment]
> during which I'll automate and test the workflow more thoroughly, unless you feel strongly otherwise?

Having this documented in the commits is already a big win and good enough for someone to pick up as an upgrade next time so I'm happy where it is at.

**Comment by robert-craig-houston** [comment]
Vtest with the architecture branch, All platforms, all targets: https://runtime-rtc.esri.com/me/my-views/view/all/job/vtest/job/3rdparty-interface/321/

**Comment by chri7325** [comment]
As a tip, you can run vTest for 3rdparty and empty out the `default` branch so you don't build all of 3rdparty, only what you have in an input for. This will give you very fast turn arounds since nothing relies on date and you can build it individually.

**Comment by robert-craig-houston** [comment]
Gotcha, thanks for the spot! 

Will be making use of that as vtest highlights build issues. I'll ping you if anything worth your attention appears!

**Comment by robert-craig-houston** [comment]
Bugfix + Reran date only vtest: https://runtime-rtc.esri.com/me/my-views/view/all/job/vtest/job/3rdparty-interface/322/
Bug: We do need some of the file access logic for accessing current timezone
Fix: Revert some ifdef logic to upstream state.

**Comment by robert-craig-houston** [comment]
Bugfix + rerun: https://runtime-rtc.esri.com/me/my-views/view/all/job/vtest/job/3rdparty-interface/323/
Bug: Multiple Define was hiding a bug where the remote API was being set to true before our defines kicked in. We then failed to fetch curl.
Fix: Modified the define logic responsible for this bug to default to true, added comment to this effect. Removed the custom logic from date.h.

**Pull request merged**
